### PR TITLE
Refactor RenderParagraph to do more work at the painting layer

### PR DESCRIPTION
**Pull request by collinjackson**

### CODE REVIEW

File: sky/packages/sky/lib/painting/paragraph_painter.dart [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import 'dart:sky' as sky;
+
+import 'package:sky/painting/text_style.dart';
+
+// This must be immutable, because we won't notice when it changes
+abstract class TextSpan {
+  sky.Node _toDOM(sky.Document owner);
+  String toString([String prefix = '']);
+
+  void _applyStyleToContainer(sky.Element container) {
```
Comment (eseidelGoogle) [inline sky/packages/sky/lib/painting/paragraph_painter.dart]: Are you attempting to make this implementation optional by adding the {}?

File: sky/packages/sky/lib/painting/paragraph_painter.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        || children.length != other.children.length)
+      return false;
+    for (int i = 0; i < children.length; ++i) {
+      if (children[i] != other.children[i])
+        return false;
+    }
+    return true;
+  }
+
+  int get hashCode {
```
Comment (eseidelGoogle) [inline sky/packages/sky/lib/painting/paragraph_painter.dart]: Quiver can do this for us.  I believe we (at least transitively) depend on quiver these days.

**Comment by eseidelGoogle** [comment]
This is great!  lgtm

File: sky/packages/sky/lib/painting/paragraph_painter.dart (hunk 1 above)
Comment (collinjackson) [inline sky/packages/sky/lib/painting/paragraph_painter.dart]: Yes. (Actually this code existed before this pull request, I just moved it around.)

File: sky/packages/sky/lib/painting/paragraph_painter.dart (hunk 2 above)
Comment (collinjackson) [inline sky/packages/sky/lib/painting/paragraph_painter.dart]: Yeah this is covered in https://github.com/domokit/sky_engine/issues/62

**Pull request merged**